### ISSUE TITLE
Historical data is missing in query results after v1.83.0

### ISSUE THREAD

**Issue by guliangsi** [VictoriaMetrics/VictoriaMetrics#3502]
**Describe the bug**
Hello,

We are having an issue after upgrading victoriametrics from v1.48.0-cluster to v1.83.1-cluster

the upgrade went well and our tests after the upgrade (including the next day) were all ok. The data history (6month) was present on the Grafana dashboards.

It seems, since today, that the data history is no longer present. The graphs all start at 12/12 8:02 p.m., which corresponds to the date of the upgrade.

Yet the retentionPeriod flag has not changed: -retentionPeriod=6
And the data seems to be present on the vmstorage disks

[xxxx@vmstorage1 big]# cd /data/victoriametrics/data/big
[xxxx@vmstorage1 big]# ll
total 20
drwxr-xr-x  5 root root  112 Dec 12 20:07 2022_06
drwxr-xr-x  6 root root  193 Dec 12 20:07 2022_07
drwxr-xr-x  8 root root 4096 Dec 12 20:07 2022_08
drwxr-xr-x  9 root root 4096 Dec 12 20:06 2022_09
drwxr-xr-x  8 root root 4096 Dec 12 20:06 2022_10
drwxr-xr-x  8 root root 4096 Dec 12 20:07 2022_11
drwxr-xr-x 14 root root 4096 Dec 19 10:36 2022_12
drwxr-xr-x  3 root root   45 Aug  1 14:55 snapshots

Is it a bug of this version? How to recover this data?
Thank you in advance for your help !

**Screenshots**
![Capture web_19-12-2022_112024_grafana-ui prod observabilite-grafana apps maif local](https://user-images.githubusercontent.com/43655090/208403603-4236edde-465c-49e9-ab31-0f9742d0dbaf.jpeg)

**Version**
vmstorage-20221110-224904-tags-v1.83.1-cluster-0-g6d984cdef

**Comment by guliangsi** [comment]
additional things that might help! The Disk sapce usage (index) graph shows that the size of the indexes on the vmstorage drastically dropped on 12/18 at 5:00 am, going from 35gb/storage to 25gb/storage.

in the logs of each vmstorage I also find this line on 12/18 at 05:00 (with different indexdb id for each vmstorage)
{"ts":"2022-12-18T04:00:00.040Z","level":"info","caller":"VictoriaMetrics/lib/storage/index_db.go:314","msg":"indexDB \"/data/victoria/metrics/indexdb/16B804EAC6BD6AA4\" has been dropped"}

![Capture web_19-12-2022_1363_grafana9-ui preprod observabilite-grafana apps maif local](https://user-images.githubusercontent.com/43655090/208423160-d501de84-cf86-4966-bb4a-9d9065e020be.jpeg)

**Comment by hwuethrich** [comment]
@guliangsi we have the same issue and are currently investigating with VM engineers. our index got dropped at the same time as yours (we have 24 month retention). just to let you know.

**Comment by Yjjko** [comment]
@hwuethrich @guliangsi 
we have the same issue too! Our index got dropped on 12/18 at 07:00 (UTC+3). In the logs we find this lines (kubernetes logs). We have 12 month retention. 

```
2022-12-18T04:00:00.001Z	info	VictoriaMetrics/lib/mergeset/table.go:270	opening table "/vm-data/indexdb/1729868653D309B8"...
2022-12-18T04:00:00.010Z	info	VictoriaMetrics/lib/mergeset/table.go:306	table "/vm-data/indexdb/1729868653D309B8" has been opened in 0.009 seconds; partsCount: 0; blocksCount: 0, itemsCount: 0; sizeBytes: 0
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:343	waiting for raw items flusher to stop on "/vm-data/indexdb/16AA0C5EB16C97F9"...
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:346	raw items flusher stopped in 0.000 seconds on "/vm-data/indexdb/16AA0C5EB16C97F9"
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:348	waiting for converters to stop on "/vm-data/indexdb/16AA0C5EB16C97F9"...
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:351	converters stopped in 0.000 seconds on "/vm-data/indexdb/16AA0C5EB16C97F9"
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:353	waiting for part mergers to stop on "/vm-data/indexdb/16AA0C5EB16C97F9"...
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:356	part mergers stopped in 0.000 seconds on "/vm-data/indexdb/16AA0C5EB16C97F9"
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:358	flushing inmemory parts to files on "/vm-data/indexdb/16AA0C5EB16C97F9"...
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:382	0 inmemory parts have been flushed to files in 0.000 seconds on "/vm-data/indexdb/16AA0C5EB16C97F9"
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:384	waiting for flush callback worker to stop on "/vm-data/indexdb/16AA0C5EB16C97F9"...
2022-12-18T04:00:00.013Z	info	VictoriaMetrics/lib/mergeset/table.go:387	flush callback worker stopped in 0.000 seconds on "/vm-data/indexdb/16AA0C5EB16C97F9"
2022-12-18T04:00:01.430Z	info	VictoriaMetrics/lib/storage/index_db.go:312	dropping indexDB "/vm-data/indexdb/16AA0C5EB16C97F9"
2022-12-18T04:00:07.399Z	info	VictoriaMetrics/lib/storage/index_db.go:314	indexDB "/vm-data/indexdb/16AA0C5EB16C97F9" has been dropped
```
also we had active merge at the same time 
![изображение](https://user-images.githubusercontent.com/29686516/208435489-c47e7219-0da3-4b59-bb4a-560c29080369.png)
![изображение](https://user-images.githubusercontent.com/29686516/208436494-16db61b0-d5f1-4afe-9370-0d87fb318b63.png)

**Comment by psalaberria002** [comment]
We are in 1.85.0, and we have also the same issue with dropped indexed from the exact same time (December 18th 5:00).

**Comment by hwuethrich** [comment]
Quote from @hagen1778 
> So far, it seems like there is unknown bug which results into currentIndex not being populated by all the active series in the current retentionPeriod. We're investigating why it could be so.

**Comment by hagen1778** [comment]
So far we think the issue was introduced in 1.83.0 and revealed itself only after the indexdb rotation. It affects the read path only, no data corruption. Rolling back to 1.82.1 and [resetting vmstorage caches](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#cache-removal) are supposed to fix the issue. We're working on the hotfix.

Please see related https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3504.

**Comment by MCPDanilovich** [comment]
> So far we think the issue was introduced in 1.83.0 and revealed itself only after the indexdb rotation. It affects the read path only, no data corruption. Rolling back to 1.82.1 and [resetting vmstorage caches](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#cache-removal) are supposed to fix the issue. We're working on the hotfix.
> 
> Please see related #3504.

Have logs to be available immediately or we have to wait some time ? I completed all steps (rollback + reset cache) and nothing changed.

**Comment by hagen1778** [comment]
Did you restart the process after the cache reset? The effect should be immediate.

**Comment by MCPDanilovich** [comment]
> Did you restart the process after the cache reset? The effect should be immediate.

I'am so sorry. It's my mistake. I have to envs and changes were applied in env where retention period to small. 
For prod env it works!

**Comment by valyala** [comment]
The commit 11bd290201e812f8b472d52cec8c6b6a8d7b48d5 should fix the root cause of the issue. The issue leads to missing data in query results for historical time series, which stop receiving new samples after the `indexdb` rotation. The `indexdb` rotation is performed once per `-retentionPeriod`. The following alerting rule can be used for detecting the issue:

```
rate(vm_missing_tsids_for_metric_id_total[5m]) > 0
```

@hagen1778 , let's add this alerting rule into [the official set of alerting rules for VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/deployment/docker/alerts.yml), in order to detect the issue quickly next time.

The issue has been introduced in [v1.83.0](https://docs.victoriametrics.com/CHANGELOG.html#v1830) in the commit 2dd93449d82ed8249e35002d6c368302bc87ec2a , so it affects all the releases between `v1.83.0` and `v1.85.1`, including these two releases.

**The issue affects only query path - the data isn't lost**. Temporary workaround is to downgrade to [v1.82.1](https://docs.victoriametrics.com/CHANGELOG.html#v1821) and [reset caches](https://docs.victoriametrics.com/#cache-removal) until the bugfix release is published.

**Comment by valyala** [comment]
The bugfix has been included in [VictoriaMetrics v1.85.2](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

@hwuethrich , @psalaberria002 , could you try [upgrading to v1.85.2](https://docs.victoriametrics.com/#how-to-upgrade-victoriametrics) and verifying whether the issue is fixed there? Do not forget [resetting caches](https://docs.victoriametrics.com/#cache-removal) before starting the upgraded VictoriaMetrics or `vmselect` nodes.

**Comment by psalaberria002** [comment]
> The bugfix has been included in [VictoriaMetrics v1.85.2](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
> 
> @hwuethrich , @psalaberria002 , could you try [upgrading to v1.85.2](https://docs.victoriametrics.com/#how-to-upgrade-victoriametrics) and verifying whether the issue is fixed there? Do not forget [resetting caches](https://docs.victoriametrics.com/#cache-removal) before starting the upgraded VictoriaMetrics or `vmselect` nodes.

We did the forward fix by upgrading to 1.85.2, and it is working 👍Thanks for the quick fix.

**Comment by mckennajones** [comment]
`v1.85.2` also resolved this issue for us (we were on `v1.84.0` before). Thanks for the super quick fix!

**Comment by valyala** [comment]
Then closing the issue as fixed.

Please note that the `v1.85.2` can trigger [another issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3515), which has been fixed in [v1.85.3](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by monteiro-renato-cog** [comment]
> The commit [11bd290](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/11bd290201e812f8b472d52cec8c6b6a8d7b48d5) should fix the root cause of the issue. The issue leads to missing data in query results for historical time series, which stop receiving new samples after the `indexdb` rotation. The `indexdb` rotation is performed once per `-retentionPeriod`. The following alerting rule can be used for detecting the issue:
> 
> ```
> rate(vm_missing_tsids_for_metric_id_total[5m]) > 0
> ```
> 
> @hagen1778 , let's add this alerting rule into [the official set of alerting rules for VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/deployment/docker/alerts.yml), in order to detect the issue quickly next time.
> 
> The issue has been introduced in [v1.83.0](https://docs.victoriametrics.com/CHANGELOG.html#v1830) in the commit [2dd9344](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/2dd93449d82ed8249e35002d6c368302bc87ec2a) , so it affects all the releases between `v1.83.0` and `v1.85.1`, including these two releases.
> 
> **The issue affects only query path - the data isn't lost**. Temporary workaround is to downgrade to [v1.82.1](https://docs.victoriametrics.com/CHANGELOG.html#v1821) and [reset caches](https://docs.victoriametrics.com/#cache-removal) until the bugfix release is published.

Hey,

Was there any alert rule created to detect this issue? I wasn't able to find it in [the official set of alerting rules for VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/deployment/docker/alerts.yml)

Thanks in advance.

**Comment by hagen1778** [comment]
@monteiro-renato-cog thanks for noticing!
Here's the PR for the new alerting rule https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3959
